File: src/main/java/com/portofino/realtrainmodunofficial/installedobject/InstalledObjectRegistry.java
package com.portofino.realtrainmodunofficial.installedobject;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.EnumMap;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class InstalledObjectRegistry {
    private static final List<InstalledObjectDefinition> ALL = new ArrayList<>();
    private static final Map<String, InstalledObjectDefinition> BY_ID = new HashMap<>();
    private static final Map<InstalledObjectCategory, List<InstalledObjectDefinition>> BY_CATEGORY =
        new EnumMap<>(InstalledObjectCategory.class);

    private InstalledObjectRegistry() {
    }

    public static void setDefinitions(List<InstalledObjectDefinition> definitions) {
        ALL.clear();
        BY_ID.clear();
        BY_CATEGORY.clear();
        for (InstalledObjectDefinition definition : definitions) {
            InstalledObjectDefinition previous = BY_ID.put(definition.getId(), definition);
            if (previous != null) {
                ALL.remove(previous);
                List<InstalledObjectDefinition> existing = BY_CATEGORY.get(previous.getCategory());
                if (existing != null) {
                    existing.remove(previous);
                }
            }
            ALL.add(definition);
            BY_CATEGORY.computeIfAbsent(definition.getCategory(), key -> new ArrayList<>()).add(definition);
        }
        ALL.sort(Comparator.comparing(InstalledObjectDefinition::getDisplayName, String.CASE_INSENSITIVE_ORDER));
        for (List<InstalledObjectDefinition> list : BY_CATEGORY.values()) {
            list.sort(Comparator.comparing(InstalledObjectDefinition::getDisplayName, String.CASE_INSENSITIVE_ORDER));
        }
    }

    public static InstalledObjectDefinition getById(String id) {
        if (id == null) return null;
        InstalledObjectDefinition found = BY_ID.get(id);
        if (found != null) return found;
        // Legacy fallback: block entities from old saves may use a different packName
        // (e.g. "RTM-Official-Assets.zip" vs the current mod JAR path).
        // Try matching by category:name, ignoring the packName segment.
        int first = id.indexOf(':');
        int last = id.lastIndexOf(':');
        if (first >= 0 && last > first) {
            String category = id.substring(0, first);
            String name = id.substring(last + 1);
            for (InstalledObjectDefinition def : ALL) {
                String defId = def.getId();
                int dFirst = defId.indexOf(':');
                int dLast = defId.lastIndexOf(':');
                if (dFirst < 0 || dLast <= dFirst) continue;
                if (defId.substring(0, dFirst).equals(category)
                        && defId.substring(dLast + 1).equals(name)) {
                    return def;
                }
            }
        }
        return null;
    }

    public static List<InstalledObjectDefinition> getByCategory(InstalledObjectCategory category) {
        return List.copyOf(BY_CATEGORY.getOrDefault(category, List.of()));
    }

    public static List<InstalledObjectDefinition> getAll() {
        return List.copyOf(ALL);
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/installedobject/InstalledObjectRegistry.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.installedobject

import java.util.EnumMap

object InstalledObjectRegistry {
    private val all: MutableList<InstalledObjectDefinition> = ArrayList()
    private val byId: MutableMap<String, InstalledObjectDefinition> = HashMap()
    private val fallbackById: MutableMap<String, InstalledObjectDefinition> = HashMap()
    private val missingIds: MutableSet<String> = HashSet()
    private val byCategory: MutableMap<InstalledObjectCategory, MutableList<InstalledObjectDefinition>> =
        EnumMap(InstalledObjectCategory::class.java)

    @JvmStatic
    fun setDefinitions(definitions: List<InstalledObjectDefinition>) {
        all.clear()
        byId.clear()
        fallbackById.clear()
        missingIds.clear()
        byCategory.clear()
        for (definition in definitions) {
            val previous = byId.put(definition.id, definition)
            if (previous != null) {
                all.remove(previous)
                byCategory[previous.category]?.remove(previous)
            }
            all.add(definition)
            byCategory.computeIfAbsent(definition.category) { ArrayList() }.add(definition)
        }
        all.sortWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.displayName })
        for (list in byCategory.values) {
            list.sortWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.displayName })
        }
    }

    @JvmStatic
    fun getById(id: String?): InstalledObjectDefinition? {
        if (id == null) {
            return null
        }
        byId[id]?.let { return it }
        fallbackById[id]?.let { return it }
        if (id in missingIds) {
            return null
        }

        val first = id.indexOf(':')
        val last = id.lastIndexOf(':')
        if (first >= 0 && last > first) {
            val category = id.substring(0, first)
            val name = id.substring(last + 1)
            for (definition in all) {
                val defId = definition.id
                val defFirst = defId.indexOf(':')
                val defLast = defId.lastIndexOf(':')
                if (defFirst < 0 || defLast <= defFirst) {
                    continue
                }
                if (defId.substring(0, defFirst) == category && defId.substring(defLast + 1) == name) {
                    fallbackById[id] = definition
                    return definition
                }
            }
        }
        missingIds.add(id)
        return null
    }

    @JvmStatic
    fun getByCategory(category: InstalledObjectCategory): List<InstalledObjectDefinition> {
        return byCategory.getOrDefault(category, emptyList()).toList()
    }

    @JvmStatic
    fun getAll(): List<InstalledObjectDefinition> = all.toList()
}
